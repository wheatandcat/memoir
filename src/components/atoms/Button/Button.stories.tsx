import React from 'react';
import { StyleSheet } from 'react-native';
import View from 'components/atoms/View';
import { mockFn } from 'storyBookUtils/index';
import Button from '.';

const Story = () => (
  <View>
    <View style={styles.inner}>
      <View p={2}>
        <Button
          title="size:sm"
          size="sm"
          width={150}
          onPress={mockFn('onPress')}
        />
      </View>
      <View p={2}>
        <Button
          title="loading & sm"
          size="sm"
          loading
          width={150}
          onPress={mockFn('onPress')}
        />
      </View>
    </View>
    <View style={styles.inner}>
      <View p={2}>
        <Button
          title="size:base"
          size="base"
          onPress={mockFn('onPress')}
          width={150}
        />
      </View>
      <View p={2}>
        <Button
          title="loading & base"
          size="base"
          loading
          onPress={mockFn('onPress')}
          width={150}
        />
      </View>
    </View>
    <View style={styles.inner}>
      <View p={2}>
        <Button
          title="size:lg"
          size="lg"
          onPress={mockFn('onPress')}
          width={150}
        />
      </View>
      <View p={2}>
        <Button
          title="loading & lg"
          size="lg"
          loading
          onPress={mockFn('onPress')}
          width={150}
        />
      </View>
    </View>

    <View p={2}>
      <Button title="disabled" disabled onPress={mockFn('onPress')} />
    </View>
  </View>
);

export default {
  title: 'atoms/Button',
  component: Story,
};

export const Default = {};

const styles = StyleSheet.create({
  inner: {
    flexDirection: 'row',
  },
});
