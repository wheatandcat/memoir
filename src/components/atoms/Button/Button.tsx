import React from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  ActivityIndicator,
  ViewStyle,
} from 'react-native';
import theme from 'config/theme';
import View from 'components/atoms/View';
import Text from 'components/atoms/Text';

type Props = {
  size?: 'sm' | 'base' | 'lg';
  width?: number | string;
  title: string;
  loading?: boolean;
  disabled?: boolean;
  onPress: () => void;
};

const Button: React.FC<Props> = (props) => {
  let configSize = theme().button.size[props.size || 'base'];

  if (!configSize) {
    configSize = theme().button.size.base;
  }

  const buttonStyle: ViewStyle = {
    paddingHorizontal: configSize.padding,
    borderRadius: configSize.borderRadius,
    height: configSize.height,
  };

  if (props.width) {
    buttonStyle.width = props.width;
  }

  if (props.loading) {
    return (
      <View style={[styles.disabledButton, buttonStyle]}>
        <ActivityIndicator size={configSize.loadingSize} />
      </View>
    );
  }

  if (props.disabled) {
    return (
      <View style={[styles.disabledButton, buttonStyle]}>
        <Text textAlign="center" size={configSize.fontSize} color="baseDark">
          {props.title}
        </Text>
      </View>
    );
  }

  return (
    <TouchableOpacity onPress={props.onPress}>
      <View style={[styles.button, buttonStyle]}>
        <Text textAlign="center" size={configSize.fontSize}>
          {props.title}
        </Text>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    backgroundColor: theme().color.primary.main,
    alignItems: 'center',
    justifyContent: 'center',
  },
  disabledButton: {
    backgroundColor: theme().color.base.main,
    alignItems: 'center',
    justifyContent: 'center',
  },
});

Button.defaultProps = {
  size: 'base',
};

export default Button;
