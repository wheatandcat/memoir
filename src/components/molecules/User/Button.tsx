import React, { memo } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import View from 'components/atoms/View';
import { User } from 'queries/api/index';
import UserImage from 'components/molecules/User/Image';

export type Props = {
  user: Pick<User, 'id' | 'image'>;
  selected: boolean;
  onAdd: (id: User['id']) => void;
  onRemove: (id: User['id']) => void;
  size?: number;
};

const Users: React.FC<Props> = (props) => {
  if (props.selected) {
    return (
      <View m={2}>
        <TouchableOpacity onPress={() => props.onRemove(props.user.id)}>
          <UserImage size={props.size} image={props.user.image} />
        </TouchableOpacity>
      </View>
    );
  }

  return (
    <View m={2} style={styles.clear}>
      <TouchableOpacity onPress={() => props.onAdd(props.user.id)}>
        <UserImage size={props.size} image={props.user.image} />
      </TouchableOpacity>
    </View>
  );
};

Users.defaultProps = {
  size: 70,
};

export default memo(Users);

const styles = StyleSheet.create({
  clear: {
    opacity: 0.4,
  },
});
