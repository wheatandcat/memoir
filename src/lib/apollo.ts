import { setContext } from '@apollo/client/link/context';
import { Alert } from 'react-native';
import { onError } from '@apollo/client/link/error';
import { ApolloClient, InMemoryCache, createHttpLink } from '@apollo/client';
import { storageKey, getItem } from 'lib/storage';
import Auth from 'lib/auth';

export const cache = new InMemoryCache();
const auth = new Auth();

type Param = {
  Authorization?: string;
  UserID?: string;
};

const makeApolloClient = async () => {
  const uri = `${process.env.API_HOST}/query`;

  const authLink = setContext(async (_, { headers }) => {
    const h = headers;
    const param: Param = {};
    const token = await auth.getIdToken();

    if (token) {
      param.Authorization = `Bearer ${token}`;
    } else {
      const userID = await getItem(storageKey.USER_ID_KEY);

      if (userID) {
        param.UserID = userID;
      }
    }

    return {
      headers: {
        ...h,
        ...param,
      },
    };
  });

  const errorLink = onError((error) => {
    console.log(error.graphQLErrors);
    if ((error.graphQLErrors || []).length > 0) {
      const graphQLErrors = error.graphQLErrors || [
        { message: 'エラー発生しました' },
      ];

      const message = graphQLErrors[0].message;
      Alert.alert('エラー', message);
    }
  });

  return new ApolloClient({
    link: errorLink.concat(authLink.concat(createHttpLink({ uri }))),
    cache: cache,
  });
};

export default makeApolloClient;
