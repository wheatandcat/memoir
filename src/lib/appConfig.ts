import firebase from 'lib/system/firebase';

export type AppConfig = {
  maintenance: boolean;
  maintenanceMessage: string;
  maintenancePeriod: string | null;
  supportVersion: string;
};

export const defaultAppConfig = () => ({
  maintenance: false,
  maintenanceMessage: '',
  maintenancePeriod: null,
  supportVersion: '1.0.0',
});

export const getAppConfig = async (
  db: firebase.firestore.Firestore
): Promise<AppConfig> => {
  const qs = await db.collection('appConfig').limit(1).get();

  const records = qs.docs.map((elem) => {
    return elem.data();
  });

  if (!records || !records[0]) {
    return defaultAppConfig();
  }

  return records[0] as AppConfig;
};
