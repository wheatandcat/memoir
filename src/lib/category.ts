import { StyleSheet, ViewStyle } from 'react-native';
import theme from 'config/theme';
import master from 'lib/master';

export const categoryBorderStyle = (category: number): ViewStyle => {
  switch (category) {
    case master.CATEGORY_1:
      return styles.category1;
    case master.CATEGORY_2:
      return styles.category2;
    case master.CATEGORY_3:
      return styles.category3;
  }

  return styles.nonCategory;
};

const styles = StyleSheet.create({
  nonCategory: {
    borderLeftColor: theme().color.background.main,
  },
  category1: {
    borderLeftColor: theme().category.color.category1,
  },
  category2: {
    borderLeftColor: theme().category.color.category4,
  },
  category3: {
    borderLeftColor: theme().category.color.category2,
  },
});
