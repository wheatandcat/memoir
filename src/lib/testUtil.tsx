import React from 'react';
import { GraphQLHandler, GraphQLRequest } from 'msw';
import {
  ApolloClient,
  ApolloProvider,
  createHttpLink,
  InMemoryCache,
} from '@apollo/client';
import { render } from '@testing-library/react-native';
import { server } from 'mocks/server';
import fetch from 'cross-fetch';

const link = createHttpLink({
  uri: 'http://localhost:8080/query',
  fetch,
  credentials: 'same-origin',
});

const client = new ApolloClient({
  link,
  uri: 'http://localhost:8080/query',
  cache: new InMemoryCache(),
});

export const testRenderer =
  (children: React.ReactNode) =>
  (responseOverride?: GraphQLHandler<GraphQLRequest<never>>) => {
    if (responseOverride) {
      server.use(responseOverride);
    }
    return render(<ApolloProvider client={client}>{children}</ApolloProvider>);
  };
