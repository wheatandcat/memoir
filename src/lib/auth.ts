import firebase from 'lib/system/firebase';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { storageKey } from 'lib/storage';
import dayjs from 'lib/dayjs';

const initFirebaseAuth = (): Promise<firebase.User | null> => {
  return new Promise((resolve) => {
    var unsubscribe = firebase.auth().onAuthStateChanged((user) => {
      // user オブジェクトを resolve
      resolve(user);

      // 登録解除
      unsubscribe();
    });
  });
};

class Auth {
  setSession = async (refresh = false) => {
    const user = await initFirebaseAuth();
    if (!user) {
      return null;
    }

    await AsyncStorage.setItem(storageKey.AUTH_UID_KEY, user.uid);

    const result = await user.getIdTokenResult(refresh);

    await AsyncStorage.setItem(storageKey.AUTH_ID_TOKEN_KEY, result.token);
    await AsyncStorage.setItem(
      storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY,
      String(result.claims.exp)
    );

    return result.token;
  };
  getIdToken = async () => {
    const idToken = await AsyncStorage.getItem(storageKey.AUTH_ID_TOKEN_KEY);
    if (!idToken) {
      return null;
    }

    const expiration = await AsyncStorage.getItem(
      storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY
    );

    if (Number(expiration) > dayjs().unix()) {
      return idToken;
    }

    return this.setSession(true);
  };
  logout = async () => {
    await firebase.auth().signOut();

    await AsyncStorage.removeItem(storageKey.AUTH_UID_KEY);
    await AsyncStorage.removeItem(storageKey.AUTH_ID_TOKEN_KEY);
    await AsyncStorage.removeItem(storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY);
  };
}

export default Auth;
