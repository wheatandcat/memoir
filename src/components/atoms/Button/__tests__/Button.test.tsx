import React from 'react';
import { testRenderer } from 'lib/testUtil';
import { screen } from '@testing-library/react-native';
import Button, { Props } from '../';

const propsData = (): Props => ({
  size: 'sm',
  width: 100,
  title: 'title',
  loading: false,
  disabled: false,
  onPress: jest.fn(),
});

describe('components/atoms/Button/Button.tsx', () => {
  beforeEach(() => {
    testRenderer(<Button {...propsData()} />)();
  });

  it('正常にrenderすること', () => {
    expect(screen.findAllByText('')).toBeTruthy();
  });
});
