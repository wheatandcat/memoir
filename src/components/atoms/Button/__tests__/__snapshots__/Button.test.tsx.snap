// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/atoms/Button/Button.tsx 正常にrenderすること 1`] = `
<TouchableOpacity
  onPress={[MockFunction]}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E3C95D",
          "justifyContent": "center",
        },
        Object {
          "borderRadius": 7,
          "height": 40,
          "paddingHorizontal": 16,
          "width": 100,
        },
      ]
    }
  >
    <Text
      color="secondary"
      fontFamily="NotoSansJP-Bold"
      fontWeight="bold"
      size="sm"
      textAlign="center"
      variants="body"
    >
      title
    </Text>
  </View>
</TouchableOpacity>
`;
